src/request.rs

```diff
@@ -0,0 +1,195 @@
+// Copyright 2023 Turing Machines
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Wrapper for `reqwest::Request` that asks for authentication if needed.
+
+use std::io::Write;
+use std::path::PathBuf;
+
+use anyhow::{bail, Context, Result};
+use reqwest::{header, Client, Method, Response, StatusCode};
+use url::Url;
+
+use crate::cli::ApiVersion;
+use crate::prompt;
+
+pub struct Request {
+    host: String,
+    ver: ApiVersion,
+    inner: reqwest::Request,
+}
+
+impl Request {
+    pub fn new(host: String, ver: ApiVersion) -> Result<Self> {
+        Self::construct_request(host, ver, Method::GET)
+    }
+
+    pub fn new_post(host: String, ver: ApiVersion) -> Result<Self> {
+        Self::construct_request(host, ver, Method::POST)
+    }
+
+    fn construct_request(host: String, ver: ApiVersion, method: reqwest::Method) -> Result<Self> {
+        let url = url_from_host(&host, ver.scheme())?;
+        let inner = reqwest::Request::new(method, url);
+
+        Ok(Self { host, ver, inner })
+    }
+
+    pub async fn send(mut self, client: &Client) -> Result<Response> {
+        self.add_auth_token(client, true).await?;
+
+        let resp = loop {
+            let req = self.clone().inner;
+            let resp = client.execute(req).await.context("HTTP request error")?;
+
+            if resp.status() == StatusCode::UNAUTHORIZED {
+                println!("Invalid token");
+                self.add_auth_token(client, false).await?;
+            } else {
+                break resp;
+            }
+        };
+
+        Ok(resp)
+    }
+
+    async fn add_auth_token(&mut self, client: &Client, use_cache: bool) -> Result<()> {
+        let token = if use_cache {
+            get_bearer_token(&self.host, self.ver, client).await?
+        } else {
+            request_token(&self.host, self.ver, client).await?
+        };
+
+        let header = format!("Bearer {}", token);
+
+        self.inner.headers_mut().insert(
+            header::AUTHORIZATION,
+            header.parse().context("convert to header value")?,
+        );
+
+        Ok(())
+    }
+
+    pub fn url(&self) -> &Url {
+        self.inner.url()
+    }
+
+    pub fn url_mut(&mut self) -> &mut Url {
+        self.inner.url_mut()
+    }
+
+    pub fn as_mut(&mut self) -> &mut reqwest::Request {
+        &mut self.inner
+    }
+
+    pub fn clone(&self) -> Self {
```

Comment:
i wonder if the wise thing to do here is to propagate the same copy interface as inner.. e.g. make it `try_copy()-> anyhow::Result<Self>`. Misaligning usually causes surprises for devs that who are not familiar with your code.